### ISSUE TITLE
Allow a way to download/save a new file

### ISSUE THREAD

**Issue by yzimet** [cugos/dropchop#11]
Perhaps [FileSaver.js](https://github.com/eligrey/FileSaver.js) ([blog post](http://updates.html5rocks.com/2011/08/Saving-generated-files-on-the-client-side)) would be suitable
>an HTML5 saveAs() FileSaver implementation

**Comment by svmatthews** [comment]
Thanks for this! Let me know if you get a chance to play around with this, otherwise I'll take a look soon.

**Comment by morganherlocker** [comment]
A common way to do this is posting to the gist.github.com api with an anonymous gist, then changing the href on your "save button" to point to the response.

**Comment by svmatthews** [comment]
This is what geojson.io does, yes? I definitely like how you can save in multiple formats with that service - are all of the formats pushed through gist or does it use different scripts for those different formats?

**Comment by svmatthews** [comment]
O_O 

Just started reading [this file](https://github.com/mapbox/geojson.io/blob/0f70ff7262377aff2d5affadc7a2f2e0a978a374/src/ui/file_bar.js#L471-L479) and Mapbox has an `npm` module for all of the things. :tada:

**Comment by svmatthews** [comment]
@aaronr how is this one coming along?

**Comment by aaronr** [comment]
Started it... will finish it soon.... just trying to finish up some work work

**Comment by svmatthews** [comment]
Oh definitely no rush - just didn't want to step on toes :mans_shoe:

**Comment by aaronr** [comment]
Got basic geojson working well... and shapefile download kind of working.  Using the shpwrite utility but it is flaky (looks like known issues with it).  So... wrapped in a try catch just like they do with geojson.io for now.  I have had it work on like 1/4 of the json I have loaded into the tool... and very few of the turf produced geoms.  Will have to look closer.  If it fails it gives a message to the end user asking them to try another format (i.e. download in geojson as that works).
Each layer is downloaded in its own file.  Pretty straight forward.
I added a dependency directory under the js dir to hold these 3rd party libs.  I also have the lint util skip them as they puke when linting.  

https://github.com/cugos/drop-n-chop/commit/5bb4c2945e1e53289ce2457b717397eeef174113

This is on the filesave branch and is up to date with master as of me writing this so will easily merge.  Test it out or give me the OK and I can merge it in.

**Comment by aaronr** [comment]
@svmatthews and @alukach ... thoughts on this?  I would love to merge this in at least as a starting point.

**Comment by alukach** [comment]
I did basic testing and it seemed to work as advertised.  This seems like
the type of that should have a couple of unit tests.  @aaronr are you up
for that?

**Comment by aaronr** [comment]
hmmm... I can try to come up with something... although I am not sure how to test the download.  I will have to see if I can catch the download package and test it.  I will look into it.

**Comment by svmatthews** [comment]
@aaronr this is really awesome. I say make a PR and we can start writing some tests!

**Comment by svmatthews** [comment]
@alukach do you think we should re-incorporate browserify before moving forward here? It would simplify future enhancements if we did it now.

**Comment by alukach** [comment]
I have no objections to bringing in browserify.

**Comment by mapsam** [comment]
@aaronr I'm running into some `shpwrite` issues as well here. Did you get these errors ever?

```
Offset is outside the bounds of the DataView
```

OR

```
Cannot read property 'filter' of undefined
```

**Comment by aaronr** [comment]
we where just wrapping the shp download in a 'try' just like they do in geojson.io:

https://github.com/cugos/dropchop/blob/master/src/js/operations/FileExecute.js#L40

https://github.com/mapbox/geojson.io/blob/gh-pages/src/ui/file_bar.js#L474

I have a feeling they do some shape normalization before the save that we are not doing so that they get a more consistent download, but I never looked closely as to why we are getting the errors on a majority of the files we try to download.

**Comment by alukach** [comment]
Seems to have been resolved by #128, closing...
